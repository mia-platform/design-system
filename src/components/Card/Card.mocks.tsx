/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { action } from '@storybook/addon-actions'

import { Hierarchy, Shape, Type } from '../Button/Button.types'
import { Button } from '../Button'
import { Icon } from '../Icon'

const { Danger } = Hierarchy
const { Circle } = Shape
const { Outlined } = Type

const actionIcon = (
  <Icon
    color="#b91200"
    name="PiTrash"
    size={16}
  />
)
export const actionButton = (
  <Button
    hierarchy={Danger}
    icon={actionIcon}
    shape={Circle}
    type={Outlined}
    onClick={action('click')}
  />
)

export const children = 'Card Content'
export const docLink = 'https://mia-platform.eu'
export const extra = <Button>{'Button'}</Button>
export const subtitle = 'Card Subtitle'
export const title = 'Card Title'
