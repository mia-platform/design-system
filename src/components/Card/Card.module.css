/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

.card {
  display: flex;
  flex-direction: column;
  background-color: var(--palette-common-white, #ffffff);
  border-radius: calc(var(--shape-border-radius-lg, 8) * 1px);
  border-width: calc(var(--shape-border-width-xs, 1) * 1px);
  padding: calc(var(--spacing-padding-sm, 8) * 1px) 0;
}

.content {
  padding: calc(var(--spacing-padding-lg, 16) * 1px) calc(var(--spacing-padding-xl, 24) * 1px);
  gap: calc(var(--spacing-gap-xl, 24) * 1px);
}

.docLink {
  display: flex;
  align-items: center;
}

.header {
  display: flex;
  justify-content: space-between;
  padding: calc(var(--spacing-padding-sm, 8) * 1px) calc(var(--spacing-padding-xl, 24) * 1px);
  gap: calc(var(--spacing-gap-xl, 24) * 1px);
}

.heading {
  display: flex;
  flex-direction: column;
  gap: calc(var(--spacing-gap-sm, 8) * 1px);
  overflow-x: hidden;

 :global(*) {
    margin: 0 !important;
  }
}

.title {
  display: flex;
  align-items: center;
  gap: calc(var(--spacing-gap-sm, 8) * 1px);
}
