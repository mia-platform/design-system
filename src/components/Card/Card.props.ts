/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

export type CardProps = {

  /**
   * The children nodes to be rendered within the card context.
   */
  children?: ReactNode,

  /**
   * The reference url for documentation of the card contents.
   * If present, a button is shown next to the title that, when clicked, opens the url in a new tab.
   */
  docLink?: string,

  /**
   * Content (typically, an action) to render in the top-right corner of the card.
   */
  extra?: ReactNode,

  /**
   * Whether the card is in a loading state (if so, a skeleton is shown).
   */
  isLoading?: boolean,

  /**
   * Subtitle of the card, which gives auxiliary information about its contents.
   */
  subtitle?: ReactNode,

  /**
   * Title of the card, which briefly conveys information about its contents.
   */
  title?: ReactNode,
}
