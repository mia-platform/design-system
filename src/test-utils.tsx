/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
