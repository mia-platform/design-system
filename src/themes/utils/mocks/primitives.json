{
  "colors": {
    "Red": {
      "300A": {
        "$type": "color",
        "$value": "#ff544f3d"
      },
      "100A": {
        "$type": "color",
        "$value": "#ff544f14"
      }
    },
    "Neutral": {
      "100A": {
        "$type": "color",
        "$value": "#00000014"
      }
    }
  },
  "sizing": {
    "none": {
      "$type": "dimension",
      "$value": 0
    }
  }
}
