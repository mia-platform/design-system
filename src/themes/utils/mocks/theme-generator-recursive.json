{
  "palette": {
    "action": {
      "alternate": {
        "danger": {
          "active": {
            "$type": "color",
            "$value": "{palette.action.alternate.danger.disabled}"
          },
          "disabled": {
            "$type": "color",
            "$value": "{colors.Neutral.100A}"
          },
          "focus": {
            "$type": "color",
            "$value": "{colors.Red.100A}"
          },
          "hover": {
            "$type": "color",
            "$value": "{palette.action.alternate.danger.active}"
          },
          "selected": {
            "$type": "color",
            "$value": "{colors.Red.300A}"
          }
        }
      }
    }
  }
}
