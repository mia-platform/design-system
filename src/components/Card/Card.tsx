/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useCallback, useMemo } from 'react'
import { Skeleton } from 'antd'

import { Shape, Type } from '../Button/Button.types'
import { BodyS } from '../Typography/BodyX/BodyS'
import { Button } from '../Button'
import { CardProps } from './Card.props'
import { H3 } from '../Typography/HX/H3'
import { Icon } from '../Icon'
import styles from './Card.module.css'
import { useTheme } from '../../hooks/useTheme'

const { Circle } = Shape
const { Ghost } = Type
const { card, content, header, heading } = styles

/**
 * UI component used to display content related to a single subject
 *
 * @returns {Card} Card component
 */
export const Card = ({
  children,
  docLink,
  extra,
  isLoading,
  subtitle,
  title,
}: CardProps): ReactElement => {
  const { palette } = useTheme()

  const docLinkIcon = useMemo(() => (
    <Icon color={palette?.action?.link?.active} name="PiBookOpen" size={16} />
  ), [palette?.action?.link?.active])

  const onClickDocLink = useCallback(() => window.open(docLink, '_blank'), [docLink])

  return (
    <div className={card}>
      <Skeleton
        active
        loading={isLoading}
        paragraph={Card.skeletonParagraph}
      >
        {(title || subtitle) && <div className={header}>
          <div className={heading}>
            {title && <div className={styles.title}>
              <H3 ellipsis={{ rows: 1, tooltip: title }}>{title}</H3>
              {docLink && <div className={styles.docLink}>
                <Button
                  icon={docLinkIcon}
                  shape={Circle}
                  type={Ghost}
                  onClick={onClickDocLink}
                />
              </div>}
            </div>}
            {subtitle && <BodyS>{subtitle}</BodyS>}
          </div>
          {extra}
        </div>}
        {children && <div className={content}>
          {children}
        </div>}
      </Skeleton>
    </div>
  )
}

Card.skeletonParagraph = {
  rows: 3,
  width: ['80%', '65%', '70%'],
}

Card.defaultProps = {
  isLoading: false,
}
