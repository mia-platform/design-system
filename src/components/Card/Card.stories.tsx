/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import type { Meta, StoryObj } from '@storybook/react'

import { actionButton, children, docLink, extra, subtitle, title } from './Card.mocks'
import { hugeData, rowKey, scroll, scrollableColumns } from '../Table/Table.mocks'
import { Card } from '.'
import { Table } from '../Table'

const meta = {
  component: Card,
  args: Card.defaultProps,
  argTypes: {
    children: { control: false },
    extra: { control: false },
  },
} satisfies Meta<typeof Card>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    ...meta.args,
    children,
    docLink,
    extra,
    subtitle,
    title,
  },
}

export const Title: Story = {
  args: { title },
}

export const TitleWithDocLink: Story = {
  args: { docLink, title },
}

export const HeaderWithSubtitle: Story = {
  args: { docLink, subtitle, title },
}

export const HeaderWithAction: Story = {
  args: { docLink, extra, title },
}

export const TextContent: Story = {
  args: { children },
}

export const ActionContent: Story = {
  args: {
    children: (
      <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between' }}>
        {children}
        {actionButton}
      </div>
    ),
  },
}

export const TableContent: Story = {
  args: {
    children: (
      <Table
        columns={scrollableColumns}
        data={hugeData}
        rowKey={rowKey}
        scroll={scroll}
      />
    ),
  },
}

export const Loading: Story = {
  args: { isLoading: true },
}
