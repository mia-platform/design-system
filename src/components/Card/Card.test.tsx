/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { children, docLink, extra, subtitle, title } from './Card.mocks'
import { render, screen } from '../../test-utils'
import { Card } from '.'

describe('Card Component', () => {
  const props = {
    docLink,
    extra,
    subtitle,
    title,
  }

  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('renders card correctly', () => {
    const { asFragment } = render(<Card {...props}>{children}</Card>)
    expect(asFragment()).toMatchSnapshot()

    expect(screen.getByRole('h3', { name: /Card Title/i })).toBeVisible()

    expect(screen.getByRole('button', { name: /PiBookOpen/i })).toBeVisible()
    expect(screen.getByRole('button', { name: /Button/i })).toBeVisible()

    expect(screen.getAllByRole('paragraph')).toHaveLength(1)
    expect(screen.getByText(/Card Subtitle/i)).toBeVisible()

    expect(screen.getByText(/Card Content/i)).toBeVisible()
  })

  test('renders header only', () => {
    const { asFragment } = render(<Card {...props} />)
    expect(asFragment()).toMatchSnapshot()

    expect(screen.getByRole('h3', { name: /Card Title/i })).toBeVisible()

    expect(screen.getByRole('button', { name: /PiBookOpen/i })).toBeVisible()
    expect(screen.getByRole('button', { name: /Button/i })).toBeVisible()

    expect(screen.getAllByRole('paragraph')).toHaveLength(1)
    expect(screen.getByText(/Card Subtitle/i)).toBeVisible()
  })

  test('renders content only', () => {
    const { asFragment } = render(<Card>{children}</Card>)
    expect(asFragment()).toMatchSnapshot()

    expect(screen.getByText(/Card Content/i)).toBeVisible()
  })

  test('renders skeleton correctly', () => {
    const { asFragment } = render(<Card isLoading={true} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
