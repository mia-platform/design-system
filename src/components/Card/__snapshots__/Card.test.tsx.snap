// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders card correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="header"
    >
      <div
        class="heading"
      >
        <div
          class="title"
        >
          <h3
            aria-label="Card Title"
            class="ant-typography"
            role="h3"
          >
            Card Title
          </h3>
          <div
            class="docLink"
          >
            <button
              class="ant-btn ant-btn-circle ant-btn-primary ant-btn-background-ghost button buttonMdIconOnly buttonGhost"
              type="button"
            >
              <svg
                alt="PiBookOpen"
                aria-label="PiBookOpen"
                color="#005dc9"
                fill="currentColor"
                height="16"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(0, 93, 201);"
                viewBox="0 0 256 256"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224,48H160a40,40,0,0,0-32,16A40,40,0,0,0,96,48H32A16,16,0,0,0,16,64V192a16,16,0,0,0,16,16H96a24,24,0,0,1,24,24,8,8,0,0,0,16,0,24,24,0,0,1,24-24h64a16,16,0,0,0,16-16V64A16,16,0,0,0,224,48ZM96,192H32V64H96a24,24,0,0,1,24,24V200A39.81,39.81,0,0,0,96,192Zm128,0H160a39.81,39.81,0,0,0-24,8V88a24,24,0,0,1,24-24h64Z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="ant-typography bodyS"
          role="paragraph"
        >
          Card Subtitle
        </div>
      </div>
      <button
        class="ant-btn ant-btn-primary button buttonMd"
        type="button"
      >
        <div
          class="buttonText"
        >
          Button
        </div>
      </button>
    </div>
    <div
      class="content"
    >
      Card Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`Card Component renders content only 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="content"
    >
      Card Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`Card Component renders header only 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="header"
    >
      <div
        class="heading"
      >
        <div
          class="title"
        >
          <h3
            aria-label="Card Title"
            class="ant-typography"
            role="h3"
          >
            Card Title
          </h3>
          <div
            class="docLink"
          >
            <button
              class="ant-btn ant-btn-circle ant-btn-primary ant-btn-background-ghost button buttonMdIconOnly buttonGhost"
              type="button"
            >
              <svg
                alt="PiBookOpen"
                aria-label="PiBookOpen"
                color="#005dc9"
                fill="currentColor"
                height="16"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(0, 93, 201);"
                viewBox="0 0 256 256"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224,48H160a40,40,0,0,0-32,16A40,40,0,0,0,96,48H32A16,16,0,0,0,16,64V192a16,16,0,0,0,16,16H96a24,24,0,0,1,24,24,8,8,0,0,0,16,0,24,24,0,0,1,24-24h64a16,16,0,0,0,16-16V64A16,16,0,0,0,224,48ZM96,192H32V64H96a24,24,0,0,1,24,24V200A39.81,39.81,0,0,0,96,192Zm128,0H160a39.81,39.81,0,0,0-24,8V88a24,24,0,0,1,24-24h64Z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="ant-typography bodyS"
          role="paragraph"
        >
          Card Subtitle
        </div>
      </div>
      <button
        class="ant-btn ant-btn-primary button buttonMd"
        type="button"
      >
        <div
          class="buttonText"
        >
          Button
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Card Component renders skeleton correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="ant-skeleton ant-skeleton-active"
    >
      <div
        class="ant-skeleton-content"
      >
        <h3
          class="ant-skeleton-title"
          style="width: 38%;"
        />
        <ul
          class="ant-skeleton-paragraph"
        >
          <li
            style="width: 80%;"
          />
          <li
            style="width: 65%;"
          />
          <li
            style="width: 70%;"
          />
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
